src/opm/output/eclipse/LoadRestart.cpp

```diff
@@ -75,79 +108,92 @@ class RestartFileView
         return this->sim_step_;
     }
 
-    operator const Opm::RestartIO::ecl_file_view_type*() const
+    template <typename ElmType>
+    bool hasKeyword(const std::string& vector) const
     {
-        return this->step_view_;
+        if (this->rst_file_ == nullptr) { return false; }
+
+        return this->vectors_
+            .at(ArrayType<ElmType>::T).count(vector) > 0;
     }
 
-    const Opm::RestartIO::ecl_kw_type* getKeyword(const char* kw) const
+    template <typename ElmType>
+    const std::vector<ElmType>&
+    getKeyword(const std::string& vector)
     {
-        namespace Load = Opm::RestartIO;
-
-        // Main grid only.  Does not handle/support LGR.
-        return Load::ecl_file_view_has_kw       (*this, kw)
-            ?  Load::ecl_file_view_iget_named_kw(*this, kw, 0)
-            :  nullptr;
+        return this->rst_file_->getRst<ElmType>(vector, this->report_step_);
     }
 
-private:
-    using RstFile = Opm::RestartIO::ert_unique_ptr<
-        Opm::RestartIO::ecl_file_type,
-        Opm::RestartIO::ecl_file_close>;
+    const std::vector<int>& intehead()
+    {
+        const auto& ihkw = std::string { "INTEHEAD" };
 
-    std::size_t                         sim_step_;
-    RstFile                             rst_file_;
-    Opm::RestartIO::ecl_file_view_type* step_view_ = nullptr;
+        if (! this->hasKeyword<int>(ihkw)) {
+            throw std::domain_error {
+                "Purported Restart File Does not Have Integer Header"
+            };
+        }
 
-    operator Opm::RestartIO::ecl_file_type*()
-    {
-        return this->rst_file_.get();
+        return this->getKeyword<int>(ihkw);
     }
 
-    operator const Opm::RestartIO::ecl_file_type*() const
-    {
-        return this->rst_file_.get();
-    }
+private:
+    using RstFile = std::unique_ptr<Opm::EclIO::ERst>;
+
+    using VectorColl = std::unordered_set<std::string>;
+    using TypedColl  = std::unordered_map<
+        Opm::EclIO::eclArrType, VectorColl, std::hash<int>
+        >;
+
+    RstFile     rst_file_;
+    int         report_step_;
+    std::size_t sim_step_;
+    TypedColl   vectors_;
 };
 
 RestartFileView::RestartFileView(const std::string& filename,
                                  const int          report_step)
-    : sim_step_(std::max(report_step - 1, 0))
-    , rst_file_(Opm::RestartIO::ecl_file_open(filename.c_str(), 0))
+    : rst_file_   { new Opm::EclIO::ERst{filename} }
+    , report_step_(report_step)
+    , sim_step_   (std::max(report_step - 1, 0))
```

Comment:
I am just working with this `-1` in current master, and I see it is carried over here in the same way. I do not understand why the `-1` is there at all - and suspect it is wrong. As a limiting example the restart file with `report_step == 0` is perfectly legitimate - although not particularly interesting to restart from.

Comments?

Reply:
> Edit: My conclusion is that we need them both.

Yes, I do believe we need them both.  We're reading inputs for the `report_step_`, but need to match those values up with `Schedule` objects from the `sim_step_` (essentially due to zero-based indexing in `Schedule` and one-based indexing of the restart files (SEQNUM).

If we ever get to a point where we build the `Schedule` and `EclipseState` objects&mdash;or at least their initial states&mdash;directly from information in restart files then we can drop the `sim_step_`.